grr/server/grr_response_server/gui/api_plugins/metadata_test.py

```diff
@@ -0,0 +1,382 @@
+  protobuf = tests_pb2.MetadataPrimitiveTypesMessage
+
+
+class MetadataRepeatedFieldMessage(rdf_structs.RDFProtoStruct):
+  protobuf = tests_pb2.MetadataRepeatedFieldMessage
+
+
+class MetadataEnumFieldMessage(rdf_structs.RDFProtoStruct):
+  protobuf = tests_pb2.MetadataEnumFieldMessage
+
+
+class MetadataTypesHierarchyRoot(rdf_structs.RDFProtoStruct):
+  protobuf = tests_pb2.MetadataTypesHierarchyRoot
+  rdf_deps = [
+    "MetadataTypesHierarchyCyclic",
+    "MetadataTypesHierarchyLeaf",
+  ]
+
+
+class MetadataTypesHierarchyCyclic(rdf_structs.RDFProtoStruct):
+  protobuf = tests_pb2.MetadataTypesHierarchyCyclic
+  rdf_deps = [
+    MetadataTypesHierarchyRoot,
+    "MetadataTypesHierarchyLeaf",
+  ]
+
+
+class MetadataTypesHierarchyLeaf(rdf_structs.RDFProtoStruct):
+  protobuf = tests_pb2.MetadataTypesHierarchyLeaf
+
+
+class DummyApiCallRouter(api_call_router.ApiCallRouter):
+  """Dummy ApiCallRouter implementation used for Metadata testing."""
+
+  @api_call_router.ArgsType(MetadataSimpleMessage)
+  @api_call_router.Http("GET", "/metadata_test/method1/<metadata_id>")
+  @api_call_router.Http("HEAD", "/metadata_test/method1/<metadata_id>")
+  @api_call_router.Http("POST", "/metadata_test/method1/<metadata_id>")
+  def Method1WithArgsType(self, args, token=None):
+    """Method 1 description."""
+
+  @api_call_router.ResultType(MetadataSimpleMessage)
+  @api_call_router.Http("GET", "/metadata_test/method2")
+  @api_call_router.Http("HEAD", "/metadata_test/method2")
+  @api_call_router.Http("POST", "/metadata_test/method2")
+  def Method2WithResultType(self, args, token=None):
+    """Method 2 description."""
+
+  @api_call_router.ArgsType(MetadataPrimitiveTypesMessage)
+  @api_call_router.ResultBinaryStream()
+  @api_call_router.Http("GET", "/metadata_test/method3")
+  def Method3PrimitiveTypes(self, args, token=None):
+    """Method 3 description."""
+
+  @api_call_router.ArgsType(MetadataRepeatedFieldMessage)
+  @api_call_router.Http("GET", "/metadata_test/method4")
+  @api_call_router.Http("POST", "/metadata_test/method4")
+  def Method4RepeatedField(self, args, token=None):
+    """Method 4 description."""
+
+  @api_call_router.ArgsType(MetadataEnumFieldMessage)
+  @api_call_router.Http("GET", "/metadata_test/method5")
+  @api_call_router.Http("POST", "/metadata_test/method5")
+  def Method5EnumField(self, args, token=None):
+    """Method 5 description."""
+
+  @api_call_router.ArgsType(MetadataTypesHierarchyRoot)
+  @api_call_router.ResultType(MetadataTypesHierarchyLeaf)
+  @api_call_router.Http("GET", "/metadata_test/method6")
+  def Method6TypeReferences(self, args, token=None):
+    """Method 6 description."""
+
+
+class ApiGetOpenApiDescriptionHandlerTest(api_test_lib.ApiCallHandlerTest):
+  """Test for ApiGetOpenApiDescriptionHandler."""
+  def setUp(self):
+    super(ApiGetOpenApiDescriptionHandlerTest, self).setUp()
+    self.router = DummyApiCallRouter()
+    self.router_methods = self.router.__class__.GetAnnotatedMethods()
```

Comment:
Please add an assertion that would check that we have all the expected methods in self.router_methods.

Reply:
I've added a separate test that asserts this. Please let me know if you want me to change it. Thank you!

Reply:
Looks good, thanks!

Reply:
Looks good!